### ISSUE TITLE
Foundation for proper grading

### ISSUE THREAD

**Issue by quick007** [CoursifyStudios/Coursify#259]
**Note**: Grade weighting will need to be created in a class settings page shared with units settings

tl;dr:

- Grade weighting
- Grades stored per student
- Grading periods (i.e. quarter, semester) weighting + creation from school
- UIs to update all of this

## Grades Setup & Flow

When a school sets up Coursify, they'll be able to choose a default set of weights for their school (i.e. summative 70% and formative 30%). Classes will be able to change this weighting too for their needs. Schools will also specify quarters/semesters, in a system of parent and child relations. In our school's case, the parent would be the semester, while the child would be 2 quarters and finals. The semester parent would even be a child of the school year. I realize that makes 0 sense, so here's a visualization to help, made in everyone's favorite drawing tool, ms paint:

![Image](https://user-images.githubusercontent.com/69743171/263433980-f46f839f-a0ce-4e5c-ba3b-947915e18490.png)

Did I discover the box and text tool halfway through? Who knows.

The point is, quarters can be nested in semesters which themselves can be nested in a school year. Also, Classes can be set to span 1 semester, or they can be set to span a school year.  

## Grading Implementation Specifics

Remember the percentages from the image above? No? You didn't look at the image because you understood what I said?

Well, now I'm going to force you to look at it because of how much of a masterpiece it is, though I'll spare your eyes the trauma of light mode this time:

![Image](https://user-images.githubusercontent.com/69743171/263434206-827e9c57-d27c-4c7e-847b-30c3991bc1d3.png)

The percentages are important because they dictate the actual weighting of grades. All of the child elements inside of a parent need to add up to 100%. This is so classes can be set up to span, say, 1 semester, without Coursify getting confused. Also, grading periods will include an optional start/end date. (start/end dates are required for nodes that will host a class, i.e. semesters or school years)

### UI Needed

This pr will also need to add a new tab on the admin page to adjust schoolwide settings like grade weighting and grading periods (full admin rewrite is coming Q2 iirc). As I mentioned in the note at the top of this post, class level grade weighting also needs to be included in the class settings page.

Lastly, when creating an assignment users need to be able to select what weight they want along with what grading period they want (autoselected current period of leaf node aka lowest depth child element with no children).

## Storing Grades

Grades will be stored across 3 locations and need to be kept in relative sync (although they don't need to update instantly). 

### Student Submissions

Submissions will always need to have the latest, up-to-date grade. This shouldn't be too complicated since that's where the grades end up when they're first entered.

These will be shown when a student/teacher is directly interacting with an assignment.

## `class_users` Table

The issue with these grades is that they're in a relatively unusable form. Imagine if we had to parse every single assignment submission every time a student looked at their grade. That would be awful! Worse, the teacher homepage would have to request all the grades for all their classes *every single time* it was loaded. Now, that might meet schoologys standards, but not ours.

The solution is to store an up-to-date set of grades in json in the `class_users` table. Because of how weights work, we might need to recalculate the grades of a user every single time something is graded, which would also be awful. I'm hoping we can avoid this somehow (please reply with a solution!).

These will be used in most places where the student looks at their own grade, like on the homepage (if enabled), their grades page, etc.

## Alongside classes

We also need to store some basic info with the classes. All we really need here is the number of students with each type of letter grade, which can probably be an array. This is for the homepage, as well as a (future) admin dashboard where a schools administration can take a detailed look at the entire school.
